media-proxy/src/mesh/conn_rdma.cc

```diff
@@ -0,0 +1,359 @@
+    for (size_t i = 0; i < capacity; ++i) {
+        void *buf = base_ptr + i * aligned_trx_sz;
+        buffer_queue.push(buf);
+    }
+
+    // Store the base memory block for cleanup
+    buffer_block = memory_block;
+
+    return Result::success;
+}
+
+// Cleanup the buffer queue
+void Rdma::cleanup_queue()
+{
+    if (buffer_block) {
+        // Free the entire memory block
+        std::free(buffer_block);
+        buffer_block = nullptr;
+    }
+    while (!buffer_queue.empty()) {
+        buffer_queue.pop();
+    }
+}
+
+/**
+ * @brief Configures the RDMA connection with the provided parameters.
+ * 
+ * Sets up the RDMA transfer size, connection kind, endpoint configuration,
+ * and queue size based on the input request. Ensures that the kind-direction
+ * pairing is valid (e.g., receiver with RX direction). Updates the connection
+ * state to "configured" upon success.
+ * 
+ * @param ctx The operation context.
+ * @param request The connection parameters including addresses and RDMA arguments.
+ * @param dev_port The device port identifier.
+ * @param dev_handle Reference to the RDMA device handle.
+ * @param kind The type of connection (receiver or transmitter).
+ * @param dir The data transfer direction (RX or TX).
+ * @return Result::success on successful configuration, or an error result if arguments are invalid.
+ */
+Result Rdma::configure(context::Context& ctx, const mcm_conn_param& request,
+                       const std::string& dev_port, libfabric_ctx *& dev_handle)
+{
+
+    trx_sz = request.payload_args.rdma_args.transfer_size;
+    _kind = kind();
+
+    std::memset(&ep_cfg, 0, sizeof(ep_cfg));
+    std::memcpy(&ep_cfg.remote_addr, &request.remote_addr, sizeof(request.remote_addr));
+    std::memcpy(&ep_cfg.local_addr, &request.local_addr, sizeof(request.local_addr));
+    
+    ep_cfg.dir = _kind == Kind::receiver ? direction::RX : direction::TX;
+
+    mDevHandle = dev_handle;
+
+    queue_size = request.payload_args.rdma_args.queue_size;
+
+    set_state(ctx, State::configured);
+    return Result::success;
+}
+
+/**
+ * @brief Establishes an RDMA connection by initializing the device, endpoint, buffer queue,
+ * and starting necessary threads for processing RDMA operations.
+ *
+ * Ensures proper cleanup and state updates on failure. Returns success if all steps
+ * (device initialization, endpoint setup, buffer queue creation, and endpoint configuration)
+ * complete successfully.
+ *
+ * @param ctx The operation context.
+ * @return Result::success on success, or an appropriate error result on failure.
+ */
+Result Rdma::on_establish(context::Context& ctx)
+{
+    if (init) {
+        log::error("RDMA device is already initialized")("state", "initialized");
+        set_state(ctx, State::active);
+        return Result::error_already_initialized;
+    }
+
+    init_buf_available();
+
+    int ret;
+    Result res;
+
+    if (!mDevHandle) {
+        ret = libfabric_dev_ops.rdma_init(&mDevHandle);
```

Comment:
I think, this code should not only allocate RDMA device with `rdma_init()`, but also deallocate it with rdma_deinit().
When? dunno, maybe when all RDMA sessions are closed, but I think it should be handled.

Reply:
For now I was calling ep_destroy upon shutdown, the initialized device was ready to be called upon again, I can add 

    if (mDevHandle) {
        libfabric_dev_ops.rdma_deinit(&mDevHandle);
    }
    
  In cleanup_resources if you think it's a better approach.

Reply:
No, I don't think adding `rdma_deinit` in `cleanup_resources()` is a good choice. 
We can have multiple concurrent RDMA connections, one connection should not deallocate whole libfabric context... unless it is sure that no other connection exists.

I guess the deallocation should be done by some other, higher abstraction that manages all connections. This will be done probably only at shutdown.

But even though it is likely that the higher level will call it, I would encapsulate `rdma_deinit` into some function here so as not to call the libfabric implementation directly and to maintain separation.

Reply:
Perhaps introducing a counter at rdma base class would be ok?
I'd only call on rdma_deinit if the active_connection counter would be 0 (after shutting down the last/or the only one).

Reply:
something like `static std::atomic<int> active_connections;`

Reply:
It would work. You can do it if you feel it is good approach. I only have 2 questions:
- Do we want to always deallocate libfabric resources when there is 0 connections? If for some reason this is done frequently, some delays will be introduced.
- Do we want connection class to be responsible for this resource? This may be the case, but an MCM agent, for example, would be more aware of when it should actually be initialized and destroyed, but I think the agent would not be necessary and always present. 

idk. I do not insist on any particular solution. I only know that if we allocate something we should clean it somewhere.

Reply:
The Agent always exists and accessible for Media Proxies. If the Agent is unavailable, the SDK won't create connections.

Reply:
fixed in https://github.com/OpenVisualCloud/Media-Communications-Mesh/pull/286/commits/13dda25882a90d3e6530502cc56dd0439b96ff95

Reply:
Deinitialization should happen sparsely, especially if there would be many connections on the same device. Possibly, measurable impact would be only in stress testing when there would be a lot single connections created and destroyed one by one as fast as possible.